src/Categories/Category/Monoidal/Interchange.agda

```diff
@@ -0,0 +1,603 @@
+                      ⟩
+
+    swapInner-unitʳ : ∀ {X Y} →
+                      [ (X ⊗₀ Y) ⊗₀ unit ⇒ (X ⊗₀ Y) ]⟨
+                        id ⊗₁ λ⇐   ⇒⟨ (X ⊗₀ Y) ⊗₀ (unit ⊗₀ unit) ⟩
+                        i⇒         ⇒⟨ (X ⊗₀ unit) ⊗₀ (Y ⊗₀ unit) ⟩
+                        ρ⇒ ⊗₁ ρ⇒
+                      ≈ ρ⇒
+                      ⟩
+
+  -- The interchange is a natural isomorphism.
+  
+  swapInner-naturalIso : ⊗ ∘F (⊗ ⁂ ⊗) ≃ ⊗ ∘F MonoidalProduct.⊗ M M
+  swapInner-naturalIso = niHelper (record
+    { η       = λ _ → i⇒
+    ; η⁻¹     = λ _ → i⇐
+    ; commute = λ _ → swapInner-natural
+    ; iso     = λ _ → swapInner.iso
+    })
+
+-- Shorthands for composing and inverting isomorphisms.
+
+open ≅ using () renaming (refl to idᵢ; sym to _⁻¹)
+infixr 9 _∘ᵢ_
+_∘ᵢ_ = λ {A B C} f g → ≅.trans {A} {B} {C} g f
+
+-- Braided monoidal categories have an interchange map.
+
+module BraidedInterchange {M : Monoidal C} (B : Braided M) where
+  open MonoidalReasoning M
+  open MonoidalUtilities M
+  open Braided B renaming (associator to α)
+  open Shorthands
+
+  -- Shorthands for braiding.
+
+  σ : ∀ {X Y} → X ⊗₀ Y ≅ Y ⊗₀ X
+  σ = braiding.FX≅GX 
+  module σ  {X Y} = _≅_ (σ {X} {Y})
+  private
+    σ⇒ = σ.from
+    σ⇐ = σ.to
+
+  -- The "four middle interchange" for braided tensor products.
+
+  swapInner : ∀ {W X Y Z} → (W ⊗₀ X) ⊗₀ (Y ⊗₀ Z) ≅ (W ⊗₀ Y) ⊗₀ (X ⊗₀ Z)
+  swapInner = α ⁻¹ ∘ᵢ idᵢ ⊗ᵢ (α ∘ᵢ σ ⊗ᵢ idᵢ ∘ᵢ α ⁻¹) ∘ᵢ α
+
+  module swapInner {W X Y Z} = _≅_ (swapInner {W} {X} {Y} {Z})
+  private
+    i⇒ = swapInner.from
+    i⇐ = swapInner.to
+
+  -- The interchange is a natural isomorphism.
+
+  swapInner-natural : ∀ {W₁ X₁ Y₁ Z₁ W₂ X₂ Y₂ Z₂}
+                      {f : W₁ ⇒ W₂} {g : X₁ ⇒ X₂} {h : Y₁ ⇒ Y₂} {i : Z₁ ⇒ Z₂} →
+                      i⇒ ∘ (f ⊗₁ g) ⊗₁ (h ⊗₁ i) ≈ (f ⊗₁ h) ⊗₁ (g ⊗₁ i) ∘ i⇒
+  swapInner-natural {f = f} {g} {h} {i} = begin
+      (α⇐ ∘ id ⊗₁ (α⇒ ∘ σ⇒ ⊗₁ id ∘ α⇐) ∘ α⇒) ∘ (f ⊗₁ g) ⊗₁ (h ⊗₁ i)
+    ≈⟨ pullʳ (pullʳ assoc-commute-from) ⟩
+      α⇐ ∘ id ⊗₁ (α⇒ ∘ σ⇒ ⊗₁ id ∘ α⇐) ∘ f ⊗₁ g ⊗₁ (h ⊗₁ i) ∘ α⇒
+    ≈⟨ refl⟩∘⟨ extendʳ (parallel id-comm-sym (begin
+          (α⇒ ∘ σ⇒ ⊗₁ id ∘ α⇐) ∘ g ⊗₁ (h ⊗₁ i)
+        ≈⟨ pullʳ (pullʳ assoc-commute-to) ⟩
+          α⇒ ∘ σ⇒ ⊗₁ id ∘ (g ⊗₁ h) ⊗₁ i ∘ α⇐
+        ≈⟨ refl⟩∘⟨ extendʳ (parallel (braiding.⇒.commute _) id-comm-sym) ⟩
+          α⇒ ∘ (h ⊗₁ g) ⊗₁ i ∘ σ⇒ ⊗₁ id ∘ α⇐
+        ≈⟨ extendʳ assoc-commute-from ⟩
+          h ⊗₁ (g ⊗₁ i) ∘ α⇒ ∘ σ⇒ ⊗₁ id ∘ α⇐
+        ∎)) ⟩
+      α⇐ ∘ f ⊗₁ h ⊗₁ (g ⊗₁ i) ∘ id ⊗₁ (α⇒ ∘ σ⇒ ⊗₁ id ∘ α⇐) ∘ α⇒
+    ≈⟨ extendʳ assoc-commute-to ⟩
+      (f ⊗₁ h) ⊗₁ (g ⊗₁ i) ∘ α⇐ ∘ id ⊗₁ (α⇒ ∘ σ⇒ ⊗₁ id ∘ α⇐) ∘ α⇒
+    ∎
+
+  swapInner-naturalIsomorphism : ⊗ ∘F (⊗ ⁂ ⊗) ≃ ⊗ ∘F MonoidalProduct.⊗ M M
+  swapInner-naturalIsomorphism = niHelper (record
+    { η       = λ _ → i⇒
+    ; η⁻¹     = λ _ → i⇐
+    ; commute = λ _ → swapInner-natural
+    ; iso     = λ _ → swapInner.iso
+    })
+
+  -- An alternative definition of the interchange that associates differently.
```

Comment:
OK yes, this deserves some extra explanation.

It's to do with symmetry. The domain `(X₁ ⊗ X₂) ⊗₀ (Y₁ ⊗ Y₂)` and codomain `(X₁ ⊗ Y₂) ⊗₀ (X₁ ⊗ Y₂)` of the interchange map are perfectly symmetric/balanced. But in order to apply the braiding to the  middle`X₂` and `Y₁`, we need to re-associate and that breaks the symmetry. We must first re-associate the whole expression in one direction and then the larger subterm in the other. This can be done in two ways, associate to the right first, then to the left, resulting in `X₁ ⊗ ((Y₂ ⊗₀ X₁) ⊗ Y₂)`, or vice versa, resulting in `(X₁ ⊗ (Y₂ ⊗₀ X₁)) ⊗ Y₂`. The choice is entirely arbitrary and results in two distinct interchange maps that behave the same way (as witnessed by `swapInner-coherent`).

Why define both? Because the proofs of some coherence laws become easier when the core of the expression is associated in one direction vs. the other. For example `swapInner-unitˡ` is easier to prove for the second definition, while `swapInner-unitʳ` is easier to prove for the first. `swapInner-assoc` uses both.

Should I include the above in the comment above the definition of `swapInner′`?

Reply:
@HuStmpHrrr: proofs using the interchange map and its coherence laws can indeed be parametrized. That's what the `HasInterchange` record is for (see `Categories.Functor.Monoidal.Tensor` for an example). But the proofs of the coherence laws themselves cannot really be parametrized over the definition of `swapInner` because they rely on the exact definition in the equational reasoning. Indeed, this is precisely what makes one definition more suitable for a given proof than the other. Hence I define both and use `swapInner-coherent` to change between them when necessary, but only one of the definitions gets exposed through `HasInterchange`. Does that make sense?

Reply:
Yes, you should include the comment above the definition of `swapInner'`.

Reply:
OK

> Yes, you should include the comment above the definition of `swapInner'`.

Done